sklearn/decomposition/_nmf.py

```diff
@@ -795,35 +896,39 @@ def _fit_multiplicative_update(X, W, H, beta_loss='frobenius',
     previous_error = error_at_init
 
     H_sum, HHt, XHt = None, None, None
+
     for n_iter in range(1, max_iter + 1):
-        # update W
-        # H_sum, HHt and XHt are saved and reused if not update_H
-        delta_W, H_sum, HHt, XHt = _multiplicative_update_w(
-            X, W, H, beta_loss, l1_reg_W, l2_reg_W, gamma,
-            H_sum, HHt, XHt, update_H)
-        W *= delta_W
-
-        # necessary for stability with beta_loss < 1
-        if beta_loss < 1:
-            W[W < np.finfo(np.float64).eps] = 0.
-
-        # update H
-        if update_H:
-            delta_H = _multiplicative_update_h(X, W, H, beta_loss, l1_reg_H,
-                                               l2_reg_H, gamma)
-            H *= delta_H
+        for i, slice in enumerate(gen_batches(n=n_samples,
+                                              batch_size=batch_size)):
+            # update W
+            # H_sum, HHt and XHt are saved and reused if not update_H
+            delta_W, H_sum, HHt, XHt = _multiplicative_update_w(
+                X[slice], W[slice], H, beta_loss, l1_reg_W, l2_reg_W,
+                gamma, H_sum, HHt, XHt, update_H)
+            W[slice] *= delta_W
+            # necessary for stability with beta_loss < 1
+            if beta_loss < 1:
+                W[slice][W[slice] < np.finfo(np.float64).eps] = 0.
 
-            # These values will be recomputed since H changed
-            H_sum, HHt, XHt = None, None, None
+            # update H
+            if update_H:
+                delta_H, A, B = _multiplicative_update_h(
+                    X[slice], W[slice], H, A, B, beta_loss,
+                    l1_reg_H, l2_reg_H, i, gamma, rho)
+                H *= delta_H
 
-            # necessary for stability with beta_loss < 1
-            if beta_loss <= 1:
-                H[H < np.finfo(np.float64).eps] = 0.
+                # These values will be recomputed since H changed
+                H_sum, HHt, XHt = None, None, None
+
+                # necessary for stability with beta_loss < 1
+                if beta_loss <= 1:
+                    H[H < np.finfo(np.float64).eps] = 0.
+
+        n_iter += i
 
         # test convergence criterion every 10 iterations
         if tol > 0 and n_iter % 10 == 0:
```

Comment:
Related to tracking number of minibatches, it would make sense to not check convergence every 10 full iterations. But maybe every 10 minibatch is too costly, idk.

Reply:
Should this be an exposed option perhaps?